File: design-patterns-java/src/main/java/co/com/patterns/builder/Computer.java
package co.com.patterns.builder;

public class Computer {
    private String cpu;
    private String ram;
    private String storage;
    private String gpu;

    public Computer() {
        this.cpu = "cpu - not defined";
        this.ram = "ram - not defined";
        this.storage = "storage - not defined";
        this.gpu = "gpu - not defined";
    }

    public void displayConfiguration() {
        String message = String.format("CPU: %s, RAM: %s, STORAGE: %s, GPU: %s", cpu, ram, storage, gpu);
        System.out.println("PC conf: \n".concat(message));
    }

    static class ComputerBuilder {
        private final Computer computer;

        public ComputerBuilder() {
            this.computer = new Computer();
        }

        public ComputerBuilder cpu(String cpu) {
            this.computer.setCpu(cpu);
            return this;
        }

        public ComputerBuilder ram(String ram) {
            this.computer.setRam(ram);
            return this;
        }

        public ComputerBuilder storage(String storage) {
            this.computer.setStorage(storage);
            return this;
        }

        public ComputerBuilder gpu(String gpu) {
            this.computer.setGpu(gpu);
            return this;
        }

        public Computer build() {
            return this.computer;
        }
    }

    public String getCpu() {
        return cpu;
    }

    public void setCpu(String cpu) {
        this.cpu = cpu;
    }

    public String getRam() {
        return ram;
    }

    public void setRam(String ram) {
        this.ram = ram;
    }

    public String getStorage() {
        return storage;
    }

    public void setStorage(String storage) {
        this.storage = storage;
    }

    public String getGpu() {
        return gpu;
    }

    public void setGpu(String gpu) {
        this.gpu = gpu;
    }
}

File: design-patterns-ts/builder/computer.ts
import { COLORS } from "../helpers/colors.ts";

class Computer {
  public cpu: string = "cpu - not defined";
  public ram: string = "ram - not defined";
  public storage: string = "storage - not defined";
  public gpu?: string = "gpu - not defined";

  displayConf(): void {
    console.log(
      `Conf Computer: \nCPU: ${this.cpu}\nRAM: ${this.ram}\nStorage: ${this.storage}\nGPU: ${this.gpu}`
    );
  }
}

class ComputerBuilder {
  private computer: Computer;

  constructor() {
    this.computer = new Computer();
  }

  setCpu(cpu: string): ComputerBuilder {
    this.computer.cpu = cpu;
    return this;
  }

  setRam(ram: string): ComputerBuilder {
    this.computer.ram = ram;
    return this;
  }

  setStorage(storage: string): ComputerBuilder {
    this.computer.storage = storage;
    return this;
  }

  setGpu(gpu: string): ComputerBuilder {
    this.computer.gpu = gpu;
    return this;
  }

  build(): Computer {
    return this.computer;
  }
}

function main() {
  const basicComputer = new ComputerBuilder()
    .setCpu("Intel 13900KF")
    .setRam("16Gb DDR6")
    .setStorage("256Gb")
    .setGpu("RTX 3050 8GB")
    .build();

  console.log("%cBasic Computer", COLORS.blue);
  basicComputer.displayConf();

  const premiunComputer = new ComputerBuilder()
    .setCpu("Ryzen 5500")
    .setRam("8Gb DDR4")
    .setStorage("SSD 480Gb")
    .setGpu("Raden 6600x 8Gb")
    .build();

  console.log("\n%cPremiun PC :>> ", COLORS.blue);
  premiunComputer.displayConf();
}

main();
